### ISSUE TITLE
SaveState as json object in MongoDb instead of String

### ISSUE THREAD

**Issue by shersh** [dapr/components-contrib#1494]
## Describe the feature
Save state in MongoDb as Object instead of string. 
If you save state using any SDK it will convert to ByteString when communicate with dapr environment and MongoDb Store component will save it just as string and you cannot use MongoDb in full performance with indexing and quering.

**Comment by shersh** [comment]
Here it will be always saving as string 
https://github.com/dapr/components-contrib/blob/7fdebda09b7177dca38137744b9240ed3b223885/state/mongodb/mongodb.go#L181

So we can just try to unmarshall as json it if true - save object, otherwise as is

**Comment by daixiang0** [comment]
https://github.com/dapr/components-contrib/blob/7fdebda09b7177dca38137744b9240ed3b223885/state/mongodb/mongodb.go#L198-L199

Do you mean this request makes mongo always handle it as string, not json?

**Comment by shersh** [comment]
> https://github.com/dapr/components-contrib/blob/7fdebda09b7177dca38137744b9240ed3b223885/state/mongodb/mongodb.go#L198-L199
> 
> Do you mean this request makes mongo always handle it as string, not json?

In your lines mongo just save value, but in case when value is string it will be saved just as string, if it will be interger as interger, if it will be object it will be saved as object 

But "value" will always string when using Dapr api, because it constructed from ByteString

**Comment by daixiang0** [comment]
Good catch, are you interested in enhancing it?

**Comment by rovangju** [comment]
Ran into this today. This should be re-opened as it is making MongoDB and the new state query api useless.
When it's persisted as a string it's not 'search-able' as it's not an object.

**Comment by shersh** [comment]
@rovangju exactly =) 

I think it can be fast way - just to trying deserialize to obj when saving, and if ok then save object instead of string.

**Comment by joshdcar** [comment]
Bumping this. Ran into this today as well and makes the utility of using the MongoDB state store much less useful for query use cases (at least via the SDK - assuming http api works as a workaround ).  Open to taking a stab at a PR for a fix (outlined pretty well by @shersh already).

**Comment by yaron2** [comment]
@joshdcar a PR would be highly appreciated 👍

**Comment by mukundansundar** [comment]
@joshdcar Have you tried setting the `contentType` metadata to `application/json` similar to code here https://github.com/dapr/java-sdk/blob/master/examples/src/main/java/io/dapr/examples/querystate/QuerySavedState.java#L77. 
Does that help?

**Comment by berndverst** [comment]
I think what we need is a case where we check that `contentType`  as @mukundansundar suggests. If and only if the content type is set as `application/json` we should try to store this as something other than string.

What needs to happen is for the string to be unmarshalled into `bson`, but only in the case where the content type was application/json.

Something like this
```golang
if contentType, ok := req.metadata['content-type']; ok && contentType == 'application/json' {
var bdoc interface{}
bson.UnmarshalJSON(data, &bdoc)
update := bson.M{"$set": bson.M{id: req.Key, value: bdoc, etag: uuid.NewString()}}
} else {
 // current approach
}
```

**Comment by joshdcar** [comment]
So Kudos to @mukundansundar for the suggestion on content type. I had tried the content type suggestion previously but while digging through the code to make the "Fix" I realized the constant was looking for "contentType" and I had the metadata key set to "Content-Type".  SaveState within API of the GRPC package was already handling data types correctly when the metadata value is set properly. 

Live and learn - On the positive side, I now feel a lot more comfortable working with and debugging component-contrib and dapr locally so I'm sure I'll find some other areas to contribute to in the near future :) My sense is the docs could use a little more clarity here. I'll take a look tomorrow and submit a PR with a suggestion - maybe save someone else a bit of time. 

Thanks for your help folks - feel free to close this issue.
